Implement CRF decode (Viterbi decode) for tensor

betterenvi opened the pull request:
Currently, Tensorflow doesn't support CRF decoding (Viterbi decoding) for tensor. Although `tf.contrib.crf.viterbi_decode` function can do CRF decoding, it can only be used at test time, since it accepts Numpy arrays as inputs.

Implementing CRF decoding for tensor benefits us a lot, as this makes our model more portable (e.g. we can freeze model, save it to a `.pb` file and then load it in `Golang` at test time).

This PR implements CRF decoding for tensor and adds test code for it. 

This PR's change is compatible with current API.

betterenvi committed:
Implement CRF decoding for tensors

betterenvi committed:
add test code for tensor version's CRF decoding

betterenvi committed:
made modifications according to pylint

tensorflow-jenkins commented:
Can one of the admins verify this patch?

betterenvi commented:
I signed it!

betterenvi committed:
add some comments for crf decode

betterenvi committed:
remove useless code

rmlarsen commented:
@drpngx do you feel like reviewing this?

drpngx commented on tensorflow/contrib/crf/python/kernel_tests/crf_test.py:
```diff
@@ -199,6 +200,87 @@ def testViterbiDecode(self):
+  def testCrfDecode(self):
+
+    def np_crf_decode(potentials, transition_params, seq_lens):
+      '''This is a function for numpy arrays.
```
Use double quotes for consistency.

Here and  below, please backtick symbols, e..g `[B, T, num_tags]`.

betterenvi replied:
Fixed.

drpngx commented on tensorflow/contrib/crf/python/kernel_tests/crf_test.py:
```diff
@@ -199,6 +200,87 @@ def testViterbiDecode(self):
+      '''This is a function for numpy arrays.
+
+      Args:
+        potentials: [B, T, num_tags]
```
We need more text

betterenvi replied:
Added.

drpngx commented on tensorflow/contrib/crf/python/kernel_tests/crf_test.py:
```diff
@@ -199,6 +200,87 @@ def testViterbiDecode(self):
+    max_seq_len = 100
+    num_tags = 4
+
+    # np.random.seed(1)
```
I would suggest leaving that in. It's easier to debug if the tests are deterministic.

betterenvi replied:
Fixed.

drpngx commented on tensorflow/contrib/crf/python/kernel_tests/crf_test.py:
```diff
@@ -199,6 +200,87 @@ def testViterbiDecode(self):
+    tag_indices_v = np.random.randint(low=0,
+                                      high=num_tags,
+                                      size=[batch_size, max_seq_len])
+    # decode using numpy version.
```
Here and below: capitalize. ("Decode").

betterenvi replied:
Fixed.

drpngx commented on tensorflow/contrib/crf/python/ops/crf.py:
```diff
@@ -43,14 +43,17 @@
     "crf_sequence_score", "crf_log_norm", "crf_log_likelihood",
-    "crf_unary_score", "crf_binary_score", "CrfForwardRnnCell", "viterbi_decode"
+    "crf_unary_score", "crf_binary_score", "CrfForwardRnnCell",
+    "viterbi_decode", "crf_decode", "CrfDecodeForwardRnnCell",
```
I think the naming scheme is a bit confusing. There's already a `viterbi_decode` that appears to be working on numpy data structures. In any case, I think we should update the top comment of the module to explain the interface better.

betterenvi replied:
Sure, I will do it soon.

betterenvi replied:
I have updated the top comment.

drpngx commented on tensorflow/contrib/crf/python/kernel_tests/crf_test.py:
```diff
@@ -199,6 +200,87 @@ def testViterbiDecode(self):
 
+  def testCrfDecode(self):
+
+    def np_crf_decode(potentials, transition_params, seq_lens):
```
Can we check against `viterbi_decode`?

betterenvi replied:
I think it's fine. If `viterbi_decode` has been tested (and is correct) and tensor version has same result with it, we know tensor version is correct too.

drpngx commented on tensorflow/contrib/crf/python/kernel_tests/crf_test.py:
```diff
@@ -199,6 +200,87 @@ def testViterbiDecode(self):
+      return pred_tags, scores
+
+
+    def calc_accuracy(pred_tags, gold_tags, seq_lens):
```
That's fine. Usually, I like to have a toy, deterministic example that can be worked out by hand. Would you mind adding that?

betterenvi replied:
Then do I need to remove current example? Or just adding another one?

betterenvi replied:
Have added a deterministic example.

drpngx reviewed (changes requested):
Nice! I haven't looked at the logic yet, but here are some comments.

betterenvi committed:
add comments at the top comment of crf module and add more comments in crf_test

betterenvi committed:
capitalize first char of first word in comments

drpngx commented:
Let me know when it's ready for review again. Thanks!

betterenvi commented:
@drpngx I had made some new changes according to your review comment. You may start review again whenever you're free. Thanks!

betterenvi committed:
replace crf_decode test code with a deterministic example

betterenvi commented:
@drpngx 

I have removed the non-deterministic example and added a deterministic one. In my test, `crf_decode` has passed this test.

By the way, `crf_decode` also has passed the previous non-deterministic example.

drpngx reviewed (approved):
Nice, thanks for adding that feature!

drpngx commented:
Jenkins, test this please.

rmlarsen commented:
Cool! Thanks for getting this reviewed so quickly @drpngx !
